--- a/contrib/avro-cmake/CMakeLists.txt
+++ b/contrib/avro-cmake/CMakeLists.txt
@@ -45,7 +45,7 @@ set_target_properties (avrocpp PROPERTIES VERSION ${AVRO_VERSION_MAJOR}.${AVRO_V

 target_include_directories(avrocpp SYSTEM PUBLIC ${AVROCPP_INCLUDE_DIR})

-target_link_libraries (avrocpp PRIVATE boost::headers_only boost::iostreams)
+target_link_libraries (avrocpp PRIVATE boost::headers_only boost::iostreams boost::filesystem)

 if (SNAPPY_INCLUDE_DIR AND SNAPPY_LIBRARY)
     target_compile_definitions (avrocpp PUBLIC SNAPPY_CODEC_AVAILABLE)
