diff -Naur /tmp/flatbuffers-1.9.0/include/flatbuffers/util.h ./include/flatbuffers/util.h
--- /tmp/flatbuffers-1.9.0/include/flatbuffers/util.h	2018-06-11 17:49:16.352763066 -0700
+++ ./include/flatbuffers/util.h	2018-06-11 17:44:00.863474932 -0700
@@ -378,7 +378,7 @@
       break;
     }
   }
-  if ((static_cast<const unsigned char>(**in) << len) & 0x80) return -1;  // Bit after leading 1's must be 0.
+  if ((static_cast<unsigned char>(**in) << len) & 0x80) return -1;  // Bit after leading 1's must be 0.
   if (!len) return *(*in)++;
   // UTF-8 encoded values with a length are between 2 and 4 bytes.
   if (len < 2 || len > 4) { return -1; }
